### ISSUE TITLE
Implement disappearing / ephemeral messages

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#1900]
### Problem

Once messages are sent in a chatroom contract, they're kinda there forever, even if they're "deleted".

It would be better if there was at least an option to send a message that could be really deleted.

### Solution

Create a new type of message under `MESSAGE_TYPES`, to allow for ephemeral messages, where the contents are stored as a separate file attachment that can be deleted (EDT: @corrideat also suggests the idea of using the KV store instead, which could also work). See related issue #1827 for storing auto-deleting attachments (that needs to be closed first).

Then update the UI to support these new message types and implement them. If designs for ephemeral messages aren't ready, then just implement the feature, but don't hook it up to the UI - or make it a simple thing like an icon (💣) next to the input field that gets toggled, and when messages are sent with this icon enabled, they are sent as ephemeral message types. And when ephemeral message types are deleted, the underlying message (in its separate file form) gets deleted.

EDIT: The kind of UI I see for this is basically a dropdown with the 💣 icon that shows these options:

- 💣 (on delete)
- 💣 1hr
- 💣 6hr
- 💣 1d
- 💣 3d
- 💣 1w
- 💣 1m
- 💣 4m

And the UI would look something like this:

[ 💣 ][ (on delete) v ]

Where the [ 💣 ] is toggleble on/off (blue background highlight when on), and the dropdown is next to it which sets the setting.

**Comment by corrideat** [comment]
I've been looking into this issue and think that going with files (as originally proposed) instead of KV (as I later suggested) is probably the simplest and most flexible solution (ETA: *).

However, something that this issue doesn't take into account is that merely deleting files may not be sufficient for a 'disappearing' feature, if copies of the file still exist. Similarly, because CIDs are used, even a 'deleted' message has some on-chain information about the file that was stored.

One way to address these issues is implementing something like the [`X3DH`](https://signal.org/docs/specifications/x3dh/) protocol for disappearing messages. Signal also has the 'double ratchet' protocol after a key is established, but we probably don't need this if we run the `X3DH` protocol for each message.

`X3DH` would require some additional keys (the `3` in the name alludes to a triple DH agreement), which could be stored in the KV store. The main advantages of using something like `X3DH` is that it's a solid and industry-standard protocol and that it provides the kind of assurances one would expect about disappearing messages, such as forward-secrecy. I still need to look up how `X3DH` works for group chats.

ETA: (*) Three reasons come to mind. KV store doesn't solve the issue of 'ephemeral' messages that have attachments, so files would still need to be used, even if as a fallback. Managing one-off KV store keys sounds complicated. Files are static while KV can be changed, which can be confusing.

**Comment by taoeffect** [comment]
It's a valid issue that you've observed. If possible I'd like to address this in the simplest way possible. So one solution to this issue that doesn't involve adding a significant amount of code to properly implement X3DH with large groups would be to create a random secret with which to encrypt the file, and encrypt that secret with the secret that we currently use, and then store it in the KV store. Then when the file is deleted we'd delete the KV-pair along with the file. This would give a high level of confidence that when the file is deleted it won't be possible to prove after the fact that whatever file was actually stored was there, satisfying the term "ephemeral".

**Comment by corrideat** [comment]
I've been thinking about this feature, and I believe that it'd be best not to implement ephemeral messages on-chain. This is because of two main reasons, a technical and a non-technical one:

1. Since ephemeral messages have a finite lifetime, they'd clutter the chain with messages that are no longer relevant. This means slower contract syncs.
2. A feature like ephemeral messages also implies some level of privacy, and writing them to a permanent chain seems counter to that.

We already have a KV store, which we can use for this purpose, although the KV store, as it's currently implemented, has the downside that anyone can write to it, possibly overwriting or deleting information written by others. It doesn't seem like this is too much of a concern for a feature like ephemeral messages, but it's a limitation of the current implementation to keep in mind.

With this, this is how ephemeral messages could work:

1. We have some kind of ephemeral message index / directory stored entirely in KV. For example, a key called `ephemeralMessages`. This is so that we can subscribe to it and listen for activity.
2. When an ephemeral message is created, some basic information is added to this index with a reference to another KV entry (let's call it `msg_${randomID}`).
3. That KV entry, `msg_${randomID}`, contains some more detailed information including decryption instructions and (optionally?) points to a file manifest.
4. The chat renders these messages interleaved with contract (i.e., on-chain) messages. They're truly disappearing in the sense that they're deleted from KV / file and they're not persisted in the chain (this is good both for privacy and performance because we don't want to fill up the chain with useless disappearing message references that no longer work)

In step 1, the index could be as simple as just a list of IDs. However, it could also include some other basic information that's useful for rendering or download purposes, such as references to when the disappearing message happened (e.g., before and after references or a timestamp), references to the disappearing message lifetime and keys that can be used for decryption (for example, the public part of an ephemeral key).

These are examples of things that could go into this index. As will be noted below, some of this information, if needed, may instead (or in addition) be in `msg_${randomID}`, depending on what makes implementing this easier.

* `msg_${randomID}` reference.
* Timestamp of creation (for ordering with on-chain messages).
* Ephemeral lifetime/expiration time.
* Sender ID.
* A flag indicating if it's a text message or contains attachments.
* The public part of the ephemeral decryption key (if per-device keys are implemented later).

Since clients can subscribe to KV events, they'd get a list of disappearing messages as they happen. In fact, the main purpose of this index is to allow this to happen, be it in the form of live updates to clients or for fetching a fresh list of messages when the app starts up.

The entry in `msg_${randomID}` would contain more detailed information about the specific message. The main reason for splitting things up is to keep the index as lean as possible. Ideally, clients would not need to subscribe to keys such as `msg_${randomID}` and would be expected to fetch them as needed. The sort of information that could be included here is the same as what was mentioned that could go in the index (i.e., lifetime, position, decryption keys, etc.), although we wouldn't want to duplicate information if we can avoid it, so these are just possibilities of what would go in there. In addition, the `msg_${randomID}` key could include the (encrypted) message payload and use files for attachments, or the entire message, including possible attachments, could be a reference to a file. This is a point that I haven't entirely decided on, but maybe, since there'll be a KV key per message, it could also include the payload and avoid additional indirection and fetches.

As mentioned in the beginning, since information would be written off-chain, these ephemeral messages are ephemeral in the sense that there is no permanent record of them. However, there is another consideration with Group Income / Chelonia, which is the lack of perfect forward secrecy without a manual key rotation. Because the entire state can be recovered by a user using a password, and this includes things in the KV store, future compromise of a password (a long-term key) can result in compromise of ephemeral messages if, for whatever reason, they were not completely deleted or are somehow recoverable. In other words, if someone (e.g., the server, a federated server, a client, anyone else) has an encrypted copy of the KV store entry, and they later learn the password of an intended recipient at the time of the message, they can use this newfound knowledge to read this ephemeral message.

Unfortunately, I don't think that there's a solution to the above problem that still lets clients fully recover the state with just a password (as the two situations are mutually exclusive). For this reason, I think that for truly ephemeral messages we need to also implement device keys in some capacity, such that long term secrets (i.e., the device keys) aren't part of the chain and can't be compromised in that way. This is a consideration that runs in parallel to the above rough description about how ephemeral messages would work, and therefore it could be implemented at a later time.

Per-device keys come with the relative downside that a new device won't be able to read the disappearing messages. I'm not sure how much of a concern this _really_ is, considering they are supposed to be short lived, and this also seems to be how most apps with this type of feature operate (meaning it wouldn't be unexpected from users). However, this can be mitigated with an online (meaning a previous device needs to be online) protocol for clients to share keys with new devices. Other than in the context of 'a new device can't read current disappearing messages', this is actually a feature, as it avoids having a single guessable key that can compromise past communication, and a necessary consequence of PFS.

Per-device keys do however come with other complexities, such as the need to manage them (e.g., the ability to revoke old keys) and potential privacy concerns with a naïve implementation (meaning that unique keys need to be derived for each user-channel pair, or at least for each user-group pair, otherwise one could use them as identifiers, which could have privacy implications).

**Comment by corrideat** [comment]
To clarify where device keys come in (if at all) and why they're optional:

We can implement the proposal above without device keys, and, in fact, without any additional encryption (since KV entries are already encrypted). Basically, we can write to the index something like `msg_1234` and `msg_1234` could include `{"payload": "Hello, World", "attachments": [], "expires": "2025-04-22T00:00:00Z"}`. It'd be 'disappearing' in that sense, and we don't clutter the main chain at all. (Or, without the 'index' idea, we could add a message type that's written to the chain that says `'msg_1234'`, with identical results). Notice again how no new encryption is used whatsoever, how things work transparently and how no device keys are involved.

The problem that device keys would solve is that if someone gets hold of `msg_1234` in its encrypted form and they later can learn a user password, or some other long-term key, they can recover the contents of the message, and they can do this at any arbitrary time in the future. In other words, the system relies on content deleted being truly deleted, which is difficult to guarantee. Device keys would therefore be used to facilitate an _additional_ layer of encryption that can provide PFS.

**Comment by taoeffect** [comment]
Just jotting down my thoughts (which may or may not be in sync with the above):

- All chain messages from now on would contain references to a key in the KV store for that message
- To load the actual message you'd look up this key
- To edit the message you'd edit the value at that key
- To delete the message you'd delete the KV pair

We'd need to make sure to upgrade the KV system (if it's not already possible) to restrict updates to these KV pairs to the author of the message only.

For auto-deleting messages, perhaps we could implement a timer mechanism / service on the server.

**Comment by taoeffect** [comment]
One more note: while working on this issue, think about whether or not this issue should be done in tandem or before/after this issue: https://github.com/okTurtles/group-income/issues/1784#issuecomment-2997479111 (for improving the chatroom contract)

**Comment by taoeffect** [comment]
Note that when storing the message in the KV store, it should be stored along with any additional info the message may have, e.g. types of reactions, etc. This will ensure that when jumping to messages (#2998) the latest version of that message, along with the latest reactions, is always displayed.

EDIT: OTOH, this would allow people to edit messages. So an alternative would be to have a different key to store reactions (editable by anyone in the chat... although that has its own obvious downside). If we go this route, we [should update the KV API to allow requesting multiple keys at the same time](https://github.com/okTurtles/libcheloniajs/issues/32) to save on round trips. We should do that anyway to grab multiple messages at once when fetching messages.
